contracts/contract_transactions.md

```diff
@@ -0,0 +1,68 @@
+[back](./contracts.md)
+## Contract Transactions
+
+Contract transactions are of three types:
+- Create
+- Call
+- Disable -- To be decided.
+
+A contract is also a normal account with a balance.
+
+
+### Create Contract Transaction
+
+Anyone can create a new contract by submitting a create contract transaction.
+
+The transaction contains:
+- The address of the contract owner (the one signing and paying for the transaction) 
+- The byte code of the contract
+- The VM to use
+- A transaction fee
+- Optionally: A initial call to the contract including name, args and gas. (Can be used to emulate init function of Ethereum/Solidity contracts)
+
+
+
+```
+{ owner           :: public_key()
+, code            :: hex_bytes()
+, vm_version      :: hex_byte()
+, fee             :: amount()
+, init            :: none() | { gas :: amount()
+                              , call_data :: hex_bytes()
+                              }
+}
+```
+
+Where a hex byte is a string of two hex digits preceded with 0x,
+and hex bytes is a string of a number of hex bytes preceded with 0x.
+
+Call data is encoded depending on the ABI of the language of the contract.
+
+The transaction has to be signed with the private key of the owner.
+
+### Contract call transaction
```

Comment:
Notes on handling of failures may be beneficial.

Cases of failures:
* gas in tx turns out not to be enough
* referred contract is disabled (maybe it was enabled when caller signed this tx)
* referred contract does not exist (maybe it was existed when caller signed this tx)
* call aborts because of bad bytecode
* call aborts/fails because some of the assumptions in the code fail (e.g. purely speculatively for examples - a referred oracle does not exist, a referred oracle has unexpected type, an account balance is not high enough for debiting)

Reply:
How are failures for spend transactions handled?
Are they signaled some how or just left in the tx-pool?

Reply:
> How are failures for spend transactions handled?
> Are they signaled some how or just left in the tx-pool?

[Current handling](https://www.pivotaltracker.com/story/show/154323788/comments/184659656) of those txs is that they are not included in the block and are left in the mempool.

Reply:
Regarding transaction failures, I think an idea was to keep such transactions in the pool, as failed transactions may get valid when a fork takes over (e.g. on a fork, an account may have sufficient funds to cover spend).

But this was intended to be coupled with some mempool pruning based on defined TTL of all transactions (how long they may live in the mempool) and max limit of transactions kept in the mempool.

_(all this is just what I remember from the initial design, not sure if this has changed)_